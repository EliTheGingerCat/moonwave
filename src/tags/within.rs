use serde::Serialize;
use std::convert::TryFrom;

use crate::{diagnostic::Diagnostic, span::Span};

#[derive(Debug, PartialEq, Serialize)]
pub struct WithinTag<'a> {
    pub name: Span<'a>,
    pub source: Span<'a>,
}

impl<'a> TryFrom<Span<'a>> for WithinTag<'a> {
    type Error = Diagnostic;

    fn try_from(span: Span<'a>) -> Result<Self, Self::Error> {
        if span.is_empty() {
            return Err(span.diagnostic("This tag has stuff after it"));
        }

        Ok(Self {
            name: span,
            source: span,
        })
    }
}

#[cfg(test)]
mod test {
    use insta::assert_yaml_snapshot;

    use super::*;

    #[test]
    fn snapshot() {
        assert_yaml_snapshot!(WithinTag::try_from(Span::dummy("hey there")));
        assert_yaml_snapshot!(WithinTag::try_from(Span::dummy("")));
    }
}
