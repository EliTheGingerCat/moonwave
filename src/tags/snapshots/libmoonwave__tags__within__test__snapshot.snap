---
source: src/tags/within.rs
expression: "WithinTag::try_from(Span::dummy(\"hey there\"))"
---
Ok:
  name: hey there
